readthedocs/projects/models.py

```diff
@@ -750,28 +774,49 @@ def has_pdf(self, version_slug=LATEST):
         path = self.get_production_media_path(
             type_='pdf', version_slug=version_slug
         )
-        storage_path = self.get_storage_path(
-            type_='pdf', version_slug=version_slug
-        )
-        return os.path.exists(path) or storage.exists(storage_path)
+        if os.path.exists(path):
+            return True
+
+        if settings.RTD_BUILD_MEDIA_STORAGE:
+            storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
+            storage_path = self.get_storage_path(
+                type_='pdf', version_slug=version_slug
+            )
+            return storage.exists(storage_path)
+
+        return False
 
     def has_epub(self, version_slug=LATEST):
         path = self.get_production_media_path(
             type_='epub', version_slug=version_slug
         )
-        storage_path = self.get_storage_path(
-            type_='epub', version_slug=version_slug
-        )
-        return os.path.exists(path) or storage.exists(storage_path)
+        if os.path.exists(path):
+            return True
+
+        if settings.RTD_BUILD_MEDIA_STORAGE:
+            storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
+            storage_path = self.get_storage_path(
+                type_='epub', version_slug=version_slug
+            )
+            return storage.exists(storage_path)
+
+        return False
 
     def has_htmlzip(self, version_slug=LATEST):
```

Comment:
nit: these 3 methods are exactly the same; the only change is `type_` value. They could be refactored to use the same chunk of code.

Not needed to be done in this PR, though.

Reply:
That seems like a worthwhile refactor and it seems useful as part of this PR.